In [657]:
import numpy as np
import pandas as pd
import seaborn as sns
from statsmodels import regression as sm
import sklearn.linear_model as lm
from sklearn import model_selection as ms
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.impute import KNNImputer
from sklearn.metrics import r2_score
from sklearn.neighbors import KNeighborsRegressor
from IPython.display import display
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.model_selection import KFold
import warnings
warnings.filterwarnings('ignore')

In [603]:
train_df = pd.read_csv(r'train.csv')
test_df = pd.read_csv(r'test.csv')

In [604]:
# train_df.info()
categorical = list(train_df.select_dtypes('object').columns)
# categorical

There are 1460 samples in the training data set and 80 features. There are 43 columns with the 'object' data type,
meaning non-numeric categorical data. These features are contained in the "categorical" list. However, notice also that
the 'MSSubClass' feature is numerical-categorical. Thus, there are actually 44 categorical features.

I will select seven non-categorical features.

In [605]:
# sns.pairplot(train_df[['SalePrice', 'LotArea', 'OverallQual', 'OverallCond', 'YearRemodAdd', 'FullBath', '1stFlrSF', '2ndFlrSF']])

The plots we care about here are in row 1 (or column 1). There appears to be a correlation between sales price and:
Overall Quality, 1st Floor Area, 2nd Floor Area, and some slight correlations with Year of Remodelling, and
Number of Full Baths.

In [606]:
# sm.linear_model.OLS()

In [607]:
MSSubClass_encoded = pd.get_dummies(train_df[['MSSubClass']].astype(str))
train_df_dropped = train_df.drop('Id', axis=1)
df_encoded = pd.get_dummies(train_df_dropped)
df_encoded = pd.concat([df_encoded, MSSubClass_encoded], axis=1).drop('MSSubClass', axis=1)
# df_encoded.info(verbose=True, null_counts=True)
# df_encoded

In [608]:
split = ms.train_test_split(df_encoded, train_size=0.8)
train_split = split[0]
test_split = split[1]
# train_split

In [609]:
# encoder = OneHotEncoder(handle_unknown="ignore", sparse=False)
# encoder.fit(train_df)

In [610]:
all_columns = df_encoded.columns
columns = df_encoded.drop(['SalePrice'], axis=1).columns
train_split.columns[train_split.isna().any()].tolist()
# test_split.columns[test_split.isna().any()].tolist()

['LotFrontage', 'MasVnrArea', 'GarageYrBlt']

Setting values to the mean or zeroes could highly skew the results of a regression model.
I will use KNN to perform multivariate imputation, filling in the above columns.

In [611]:
# Training the inputer with train split
imputer_train = KNNImputer(n_neighbors=15, weights="uniform")
imputer_train.fit(train_split)
train_split = pd.DataFrame(imputer_train.fit_transform(df_encoded), columns = all_columns)
test_split = pd.DataFrame(imputer_train.fit_transform(test_split), columns = all_columns)

In [612]:
# training the normalizer with train split
normalize_train = StandardScaler().fit(train_split.drop(['SalePrice'], axis=1))

train_norm = normalize_train.transform(train_split.drop(['SalePrice'], axis=1))
test_norm = normalize_train.transform(test_split.drop(['SalePrice'], axis=1))

train_norm = pd.DataFrame(train_norm, columns = columns)
test_norm = pd.DataFrame(test_norm, columns = columns)
train_norm

LotFrontage   LotArea  OverallQual  OverallCond  YearBuilt  \
0       -0.247803 -0.207142     0.651479    -0.517200   1.050994   
1        0.412273 -0.091886    -0.071836     2.179628   0.156734   
2       -0.115788  0.073480     0.651479    -0.517200   0.984752   
3       -0.467828 -0.096897     0.651479    -0.517200  -1.863632   
4        0.588294  0.375148     1.374795    -0.517200   0.951632   
...           ...       ...          ...          ...        ...   
1455    -0.379818 -0.260560    -0.071836    -0.517200   0.918511   
1456     0.632299  0.266407    -0.071836     0.381743   0.222975   
1457    -0.203798 -0.147810     0.651479     3.078570  -1.002492   
1458    -0.115788 -0.080160    -0.795151     0.381743  -0.704406   
1459     0.192248 -0.058112    -0.795151     0.381743  -0.207594   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  ...  \
0         0.878668    0.508709    0.575425   -0.288653  -0.944591  ...   
1        -0.429577   -0.575571    1.171992   -0.288653  -0.641228  ...   
2         0.830215    0.320620    0.092907   -0.288653  -0.301643  ...   
3        -0.720298   -0.575571   -0.499274   -0.288653  -0.061670  ...   
4         0.733308    1.360644    0.463568   -0.288653  -0.174865  ...   
...            ...         ...         ...         ...        ...  ...   
1455      0.733308   -0.575571   -0.973018   -0.288653   0.873321  ...   
1456      0.151865    0.082742    0.759659    0.722112   0.049262  ...   
1457      1.024029   -0.575571   -0.369871   -0.288653   0.701265  ...   
1458      0.539493   -0.575571   -0.865548    6.092188  -1.284176  ...   
1459     -0.962566   -0.575571    0.847389    1.509640  -0.976285  ...   

      MSSubClass_30  MSSubClass_40  MSSubClass_45  MSSubClass_50  \
0         -0.222721      -0.052414      -0.091035      -0.330791   
1         -0.222721      -0.052414      -0.091035      -0.330791   
2         -0.222721      -0.052414      -0.091035      -0.330791   
3         -0.222721      -0.052414      -0.091035      -0.330791   
4         -0.222721      -0.052414      -0.091035      -0.330791   
...             ...            ...            ...            ...   
1455      -0.222721      -0.052414      -0.091035      -0.330791   
1456      -0.222721      -0.052414      -0.091035      -0.330791   
1457      -0.222721      -0.052414      -0.091035      -0.330791   
1458      -0.222721      -0.052414      -0.091035      -0.330791   
1459      -0.222721      -0.052414      -0.091035      -0.330791   

      MSSubClass_60  MSSubClass_70  MSSubClass_75  MSSubClass_80  \
0          1.970518      -0.207020      -0.105263      -0.203395   
1         -0.507481      -0.207020      -0.105263      -0.203395   
2          1.970518      -0.207020      -0.105263      -0.203395   
3         -0.507481       4.830459      -0.105263      -0.203395   
4          1.970518      -0.207020      -0.105263      -0.203395   
...             ...            ...            ...            ...   
1455       1.970518      -0.207020      -0.105263      -0.203395   
1456      -0.507481      -0.207020      -0.105263      -0.203395   
1457      -0.507481       4.830459      -0.105263      -0.203395   
1458      -0.507481      -0.207020      -0.105263      -0.203395   
1459      -0.507481      -0.207020      -0.105263      -0.203395   

      MSSubClass_85  MSSubClass_90  
0         -0.117851      -0.192177  
1         -0.117851      -0.192177  
2         -0.117851      -0.192177  
3         -0.117851      -0.192177  
4         -0.117851      -0.192177  
...             ...            ...  
1455      -0.117851      -0.192177  
1456      -0.117851      -0.192177  
1457      -0.117851      -0.192177  
1458      -0.117851      -0.192177  
1459      -0.117851      -0.192177  

[1460 rows x 302 columns]

In [660]:
X = train_norm
Y = train_split['SalePrice']
X_test = test_norm
Y_test = test_split['SalePrice']

X_non_norm = train_split.drop(['SalePrice'], axis=1)
X_test_non_norm = test_split.drop(['SalePrice'], axis=1)

In [665]:
lin_reg = lm.LinearRegression().fit(X, Y)
lin_pred = lin_reg.predict(X_test)
# lin_reg.score(test_norm.drop(['SalePrice'], axis=1), test_norm['SalePrice'])
r2_score(Y_test, lin_pred)
np.shape(lin_reg.coef_)

(302,)

In [615]:
knn_reg = KNeighborsRegressor()
param_grid = {'n_neighbors': np.arange(1, 15)}
knn_grid_cv = ms.GridSearchCV(knn_reg, param_grid, cv=10)
knn_grid_cv.fit(X, Y)
print(knn_grid_cv.best_params_)
print(knn_grid_cv.best_score_)
display(pd.DataFrame(knn_grid_cv.cv_results_))

{'n_neighbors': 11}
0.7405634361447293


mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0        0.003501      0.000500         0.005201        0.000400   
1        0.003801      0.001470         0.005801        0.001833   
2        0.003101      0.000300         0.005501        0.000500   
3        0.003701      0.000900         0.009402        0.008250   
4        0.003601      0.000490         0.007701        0.005480   
5        0.002704      0.000461         0.005498        0.000504   
6        0.002701      0.000459         0.005501        0.000671   
7        0.002801      0.000400         0.005401        0.000490   
8        0.002901      0.000300         0.005601        0.000664   
9        0.002901      0.000539         0.005401        0.000664   
10       0.002801      0.000400         0.005301        0.000640   
11       0.002701      0.000459         0.005601        0.000663   
12       0.002701      0.000458         0.005501        0.000500   
13       0.002701      0.000458         0.005201        0.000400   

   param_n_neighbors               params  split0_test_score  \
0                  1   {'n_neighbors': 1}           0.704337   
1                  2   {'n_neighbors': 2}           0.760294   
2                  3   {'n_neighbors': 3}           0.791678   
3                  4   {'n_neighbors': 4}           0.779533   
4                  5   {'n_neighbors': 5}           0.767779   
5                  6   {'n_neighbors': 6}           0.773894   
6                  7   {'n_neighbors': 7}           0.770276   
7                  8   {'n_neighbors': 8}           0.810577   
8                  9   {'n_neighbors': 9}           0.804290   
9                 10  {'n_neighbors': 10}           0.799760   
10                11  {'n_neighbors': 11}           0.802694   
11                12  {'n_neighbors': 12}           0.798888   
12                13  {'n_neighbors': 13}           0.794569   
13                14  {'n_neighbors': 14}           0.796144   

    split1_test_score  split2_test_score  split3_test_score  \
0            0.288652           0.673225           0.668503   
1            0.597509           0.785232           0.744155   
2            0.680672           0.789750           0.732312   
3            0.716608           0.789691           0.720356   
4            0.721353           0.793127           0.690670   
5            0.783611           0.795534           0.700564   
6            0.769669           0.790530           0.676201   
7            0.764780           0.791389           0.670147   
8            0.769589           0.791150           0.670639   
9            0.760243           0.791905           0.686254   
10           0.752803           0.796325           0.691095   
11           0.747025           0.793641           0.695943   
12           0.738115           0.786389           0.686157   
13           0.736029           0.784709           0.683200   

    split4_test_score  split5_test_score  split6_test_score  \
0            0.728969           0.596297           0.750835   
1            0.724942           0.711895           0.783782   
2            0.732995           0.743530           0.791826   
3            0.731071           0.745167           0.763657   
4            0.733361           0.774366           0.749560   
5            0.712149           0.785216           0.754382   
6            0.700273           0.789617           0.766473   
7            0.695494           0.779921           0.763272   
8            0.694627           0.787157           0.760536   
9            0.700491           0.779018           0.766560   
10           0.704504           0.777661           0.775733   
11           0.702116           0.763405           0.778144   
12           0.700082           0.760412           0.777863   
13           0.697719           0.756924           0.775142   

    split7_test_score  split8_test_score  split9_test_score  mean_test_score  \
0            0.632051           0.345203          

It appears that 11 neighbours gives us the best score. However, given the standard error of about 0.07, a score of at least
0.67 performs similarly. Thus, the "rule-of-thumb" best selection could be argued to be k=2 KNN with a score of 0.7014

In [616]:
ridge_reg = lm.RidgeCV(alphas=np.linspace(0.001,1000,30), cv=10)
ridge_reg.fit(X, Y)
print(ridge_reg.best_score_)
print(ridge_reg.alpha_)

0.8505264801075464
758.6209310344827


In [617]:
ridge_reg = lm.Ridge()
param_grid = {'alpha': np.linspace(0.001,10000,100)}
ridge_grid_cv = ms.GridSearchCV(ridge_reg, param_grid, cv=10)
ridge_grid_cv.fit(X, Y)
print(ridge_grid_cv.best_params_)
print(ridge_grid_cv.best_score_)
print(ridge_grid_cv.best_index_)
ridge_cv_table = pd.DataFrame(ridge_grid_cv.cv_results_)
ridge_cv_table

{'alpha': 707.0716363636363}
0.8505097456712031
7


mean_fit_time  std_fit_time  mean_score_time  std_score_time  param_alpha  \
0        0.007701      0.000459         0.001901        0.000300        0.001   
1        0.007801      0.000400         0.001901        0.000300   101.011091   
2        0.007701      0.000459         0.001601        0.000490   202.021182   
3        0.007952      0.000351         0.001901        0.000300   303.031273   
4        0.007801      0.000400         0.001901        0.000300   404.041364   
..            ...           ...              ...             ...          ...   
95       0.007802      0.000400         0.001701        0.000458  9595.959636   
96       0.009001      0.003034         0.001801        0.000400  9696.969727   
97       0.007902      0.000300         0.001801        0.000400  9797.979818   
98       0.007702      0.000641         0.001801        0.000400  9898.989909   
99       0.008001      0.000633         0.001801        0.000400      10000.0   

                           params  split0_test_score  split1_test_score  \
0                {'alpha': 0.001}           0.882797           0.824154   
1   {'alpha': 101.01109090909091}           0.908442           0.841937   
2   {'alpha': 202.02118181818182}           0.915468           0.853634   
3    {'alpha': 303.0312727272727}           0.918958           0.861450   
4    {'alpha': 404.0413636363636}           0.920918           0.866878   
..                            ...                ...                ...   
95   {'alpha': 9595.959636363636}           0.807544           0.735987   
96   {'alpha': 9696.969727272728}           0.806138           0.734420   
97   {'alpha': 9797.979818181819}           0.804734           0.732859   
98    {'alpha': 9898.98990909091}           0.803333           0.731305   
99             {'alpha': 10000.0}           0.801934           0.729756   

    split2_test_score  split3_test_score  split4_test_score  \
0            0.912537           0.751018           0.900222   
1            0.917140           0.771243           0.903549   
2            0.917034           0.778262           0.898969   
3            0.916649           0.781929           0.894194   
4            0.916232           0.784097           0.889546   
..                ...                ...                ...   
95           0.783840           0.687069           0.672809   
96           0.782338           0.685824           0.671181   
97           0.780840           0.684582           0.669561   
98           0.779346           0.683342           0.667949   
99           0.777856           0.682105           0.666346   

    split5_test_score  split6_test_score  split7_test_score  \
0            0.589256           0.886591           0.895929   
1            0.778636           0.893137           0.898473   
2            0.815121           0.893655           0.897922   
3            0.834656           0.892960           0.897239   
4            0.846311           0.891740           0.896477   
..                ...                ...                ...   
95           0.715143           0.735236           0.762134   
96           0.713599           0.733770           0.760684   
97           0.712062           0.732308           0.759239   
98           0.710532           0.730852           0.757797   
99           0.709008           0.729401           0.756360   

    split8_test_score  split9_test_score  mean_test_score  std_test_score  \
0            0.420825          -1.254833         0.580850        0.630833   
1            0.503665           0.899134         0.831536        0.120546   
2            0.536410           0.897083         0.840356        0.110173   
3            0.556682           0.895679         0.845040        0.103973   
4            0.570754           0.894672         0.847762        0.099723   
..                ...                ...              ...             ...   
95           0.560058           0.768608         0.722843        0.067231   


In [618]:
accept = ridge_grid_cv.best_score_ - ridge_cv_table['std_test_score'].iloc[ridge_grid_cv.best_index_]
print(accept)
print(ridge_cv_table[(ridge_cv_table['mean_test_score'] <= accept) & (ridge_cv_table.index > ridge_grid_cv.best_index_)].iloc[0])
ridge_cv_table[['param_alpha', 'mean_test_score']].iloc[71]

0.7584074107463838
mean_fit_time                            0.007701
std_fit_time                             0.000458
mean_score_time                          0.001801
std_score_time                             0.0004
param_alpha                           7272.727545
params               {'alpha': 7272.727545454545}
split0_test_score                        0.840509
split1_test_score                        0.773738
split2_test_score                        0.819431
split3_test_score                        0.716242
split4_test_score                        0.712863
split5_test_score                        0.752594
split6_test_score                         0.77045
split7_test_score                        0.796494
split8_test_score                        0.584314
split9_test_score                        0.802137
mean_test_score                          0.756877
std_test_score                           0.069555
rank_test_score                                72
Name: 72, dtype: object


param_alpha        7171.717455
mean_test_score       0.758408
Name: 71, dtype: object

Our best score here is with lambda or alpha ~ 707.07 with 0.8505. The standard error is 0.0921. This means that our
simplest model (higher lambda) is at index 71 (the lambda within our standard error threshold). This is lambda ~ 7172
and CV score = 0.7584

In [666]:
lasso_reg = lm.Lasso()
param_grid = {'alpha': np.linspace(0.001,20000,50)}
lasso_grid_cv = ms.GridSearchCV(lasso_reg, param_grid, cv=10)
lasso_grid_cv.fit(X, Y)
print(lasso_grid_cv.best_params_)
print(lasso_grid_cv.best_score_)
print(lasso_grid_cv.best_index_)
lasso_cv_table = pd.DataFrame(lasso_grid_cv.cv_results_)
# lasso_cv_table

{'alpha': 1632.6539795918368}
0.8443541296515326
4


In [620]:
accept = lasso_grid_cv.best_score_ - lasso_cv_table['std_test_score'].iloc[lasso_grid_cv.best_index_]
print(accept)
print(lasso_cv_table[(lasso_cv_table['mean_test_score'] <= accept) & (lasso_cv_table.index > lasso_grid_cv.best_index_)].iloc[0])
lasso_cv_table[['param_alpha', 'mean_test_score']].iloc[30]

0.7457931992421065
mean_fit_time                             0.006352
std_fit_time                               0.00045
mean_score_time                             0.0017
std_score_time                            0.000458
param_alpha                           12653.061592
params               {'alpha': 12653.061591836735}
split0_test_score                          0.80166
split1_test_score                         0.771215
split2_test_score                         0.805111
split3_test_score                         0.707488
split4_test_score                         0.706631
split5_test_score                         0.715163
split6_test_score                         0.764257
split7_test_score                         0.766628
split8_test_score                         0.596848
split9_test_score                         0.796461
mean_test_score                           0.743146
std_test_score                            0.060566
rank_test_score                                 32
Name: 31, dt

param_alpha        12244.898347
mean_test_score         0.74699
Name: 30, dtype: object

Our best score here is with lambda or alpha ~ 1632.65 with 0.8444. The standard error is 0.0986. This means that our
simplest model (higher lambda) is at index 30 (the lambda within our standard error threshold).
This is lambda ~ 12245 and CV score = 0.7470

In [661]:
lin_reg_cv = lm.LinearRegression()
cv_scores = cross_val_score(lin_reg_cv, X_non_norm, Y, cv=10)
print(np.mean(cv_scores))
print(np.std(cv_scores))
print(cv_scores)

0.5740162171877558
0.6390221064393009
[ 0.86387649  0.80127488  0.91110944  0.74827339  0.90051782  0.62523891
  0.88577184  0.89588169  0.39476086 -1.28654315]


In [645]:
v = np.linalg.inv(X.T.dot(X).values)

0.012526487497190182

In [677]:
def sse_calc(y, y_hat):
    return np.sum(np.square(y - y_hat))

In [705]:
def z_calc(beta, n, k, v_j, sse):
    sigma = np.sqrt(1/(n-k-1) * sse)
    return beta/(sigma * np.sqrt(v_j))

In [727]:
kf = KFold(n_splits=10)
kf.split(X)
n_fold = 1
z_all = np.array([])

for train_index, test_index in kf.split(X):
    z_row = []
    train_fold_X = X_non_norm.iloc[train_index]
    train_fold_Y = Y.iloc[train_index]

    test_fold_X = X_non_norm.iloc[test_index]
    test_fold_Y = Y.iloc[test_index]

    lin_reg_fold = lm.LinearRegression().fit(train_fold_X, train_fold_Y)
    pred_fold = lin_reg_fold.predict(test_fold_X)

    n = len(train_fold_X) ### should be test length
    k = len(train_fold_X.columns)
    xtx = train_fold_X.T.dot(train_fold_X).values
    v = np.linalg.pinv(xtx)

    sse = sse_calc(test_fold_Y, pred_fold)
    beta = lin_reg_fold.coef_

    for j in range(len(beta)):
        v_j = v[j,j]
        b_j = beta[j]
        z_j = z_calc(b_j, n, k, v_j, sse)
        z_row.append(z_j)
    print(z_row)
    np.append(z_all, z_row, axis=0)
    n_fold += 1
    print('new fold')


[3.5943244039084656, 11.814971773859796, 14.467882804805752, 15.261487799492066, 8.783628147756609, 4.109360670806938, 11.292742991048703, 12.649267686244585, 4.183680979945812, -1.8165130487548269, 14.01496564106805, 6.219888232846883, 15.521999591208104, -5.18020503375036, 9.065607514613177, 2.854176524440679, 0.6633823117719051, 3.5883672848979096, 0.03519511752294641, -5.3635924950890495, -3.867437367700414, 4.302801020944002, 7.163113328977122, -0.5860223992097517, 3.5864852821347712, 5.199482948394159, 6.701735001586816, -0.930185676691969, 0.18915002273455392, 2.1419065203564425, 5.808111646762562, 6.638607786636736, -0.3349824415586022, -4.257257109377372, -2.755082982581652, -0.8872625852637858, 0.7661016569300573, -0.003398547553253966, 0.22132100197711396, -0.061588027510880015, -0.31824632757009025, 0.3174181929009661, 0.7804681599663554, 1.4351912515343528, -0.1486514466731314, 0.12087814926913781, 0.03261806232080963, -0.005360031082204965, -0.09861224803126577, 0.3760346

In [724]:
print(np.shape(z_all))
























(0,)
